File: .zo/scripts/python/setup-design.py
#!/usr/bin/env python3
"""
setup-design.py
Initialize design documentation for a feature or global design system.

Usage:
    ./setup-design.py [--json] [--global] [feature-dir]

Options:
    --json    Output results in JSON format
    --global  Setup global design system (.zo/design-system.md)
    --help    Show this help message

Examples:
    ./setup-design.py                    # Auto-detect from current branch
    ./setup-design.py specs/001-feature  # Use specific feature directory
    ./setup-design.py --global           # Setup global design system

Output:
    JSON object with:
    - MODE: 'global' or 'feature'
    - DESIGN_FILE: Path to the design file
    - FEATURE_SPEC: Path to feature spec (empty in global mode)
    - FEATURE_DIR: Path to feature directory (repo root in global mode)
    - FEATURE_NAME: Feature name or 'global'
"""

import argparse
import json
import os
import re
import sys
from pathlib import Path

# Add parent directory to path to import common module
script_dir = Path(__file__).parent.resolve()
sys.path.insert(0, str(script_dir))

from common import (
    get_repo_root,
    get_current_branch,
    find_feature_dir_by_prefix,
    check_file_exists
)


def parse_args():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description='Initialize design documentation for a feature or global design system',
        add_help=False  # We'll handle --help manually
    )
    
    parser.add_argument(
        '--json',
        action='store_true',
        help='Output results in JSON format'
    )
    
    parser.add_argument(
        '--global',
        '-g',
        action='store_true',
        dest='global_mode',
        help='Setup global design system (.zo/design-system.md)'
    )
    
    parser.add_argument(
        '--help',
        '-h',
        action='store_true',
        help='Show this help message'
    )
    
    parser.add_argument(
        'feature_dir',
        nargs='?',
        default='',
        help='Feature directory (optional)'
    )
    
    return parser.parse_args()


def setup_global_design(repo_root: str, json_mode: bool) -> dict:
    """
    Setup global design system.
    
    Args:
        repo_root: Repository root path
        json_mode: Whether to output JSON
        
    Returns:
        Dictionary with result information
    """
    design_file = os.path.join(repo_root, '.zo', 'design-system.md')
    feature_spec = ''
    feature_dir = repo_root
    feature_name = 'global'
    
    # Use global design system template if available
    template_path = os.path.join(repo_root, '.zo', 'templates', 'design-system-template.md')
    
    if not check_file_exists(design_file) and check_file_exists(template_path):
        # Read template
        with open(template_path, 'r') as f:
            content = f.read()
        
        # Write to design file
        with open(design_file, 'w') as f:
            f.write(content)
    
    return {
        'MODE': 'global',
        'DESIGN_FILE': design_file,
        'FEATURE_SPEC': feature_spec,
        'FEATURE_DIR': feature_dir,
        'FEATURE_NAME': feature_name
    }


def setup_feature_design(repo_root: str, feature_arg: str, json_mode: bool) -> dict:
    """
    Setup feature design.
    
    Args:
        repo_root: Repository root path
        feature_arg: Feature directory argument (if provided)
        json_mode: Whether to output JSON
        
    Returns:
        Dictionary with result information
    """
    current_branch = get_current_branch()
    
    # Handle optional feature directory argument
    if feature_arg:
        # User provided a feature directory
        if os.path.isdir(feature_arg):
            # Absolute or relative path provided
            feature_dir = os.path.abspath(feature_arg)
            feature_spec = os.path.join(feature_dir, 'spec.md')
            feature_name = os.path.basename(feature_dir)
        elif os.path.isdir(os.path.join(repo_root, feature_arg)):
            # Path relative to repo root
            feature_dir = os.path.abspath(os.path.join(repo_root, feature_arg))
            feature_spec = os.path.join(feature_dir, 'spec.md')
            feature_name = os.path.basename(feature_dir)
        else:
            print(f"Error: Directory '{feature_arg}' not found.", file=sys.stderr)
            sys.exit(1)
    else:
        # Auto-detect context from current branch
        if re.match(r'^\d{3}-', current_branch):
            # Branch name is the feature name
            feature_dir = find_feature_dir_by_prefix(repo_root, current_branch)
            feature_spec = os.path.join(feature_dir, 'spec.md')
            feature_name = current_branch
        else:
            # Fallback to current directory if valid structure
            if check_file_exists('spec.md'):
                feature_dir = os.getcwd()
                feature_spec = os.path.join(feature_dir, 'spec.md')
                feature_name = os.path.basename(feature_dir)
            else:
                print(
                    "Error: Could not determine feature context. "
                    "Run inside a feature branch or specify directory.",
                    file=sys.stderr
                )
                sys.exit(1)
    
    # Define output file for feature design
    design_file = os.path.join(feature_dir, 'design.md')
    
    # Use feature design template if available
    template_path = os.path.join(repo_root, '.zo', 'templates', 'design-template.md')
    
    if not check_file_exists(design_file) and check_file_exists(template_path):
        # Read template
        with open(template_path, 'r') as f:
            content = f.read()
        
        # Replace basic placeholders
        content = content.replace('[FEATURE NAME]', feature_name)
        
        # Write to design file
        with open(design_file, 'w') as f:
            f.write(content)
    
    return {
        'MODE': 'feature',
        'DESIGN_FILE': design_file,
        'FEATURE_SPEC': feature_spec,
        'FEATURE_DIR': feature_dir,
        'FEATURE_NAME': feature_name
    }


def main():
    """Main entry point."""
    args = parse_args()
    
    # Handle help
    if args.help:
        print(__doc__)
        sys.exit(0)
    
    # Get repository root
    repo_root = get_repo_root()
    
    # Handle global vs feature mode
    if args.global_mode:
        result = setup_global_design(repo_root, args.json)
    else:
        result = setup_feature_design(repo_root, args.feature_dir, args.json)
    
    # Output results
    if args.json:
        print(json.dumps(result))
    else:
        if result['MODE'] == 'global':
            print(f"MODE: {result['MODE']}")
            print(f"DESIGN_FILE: {result['DESIGN_FILE']}")
        else:
            print(f"MODE: {result['MODE']}")
            print(f"DESIGN_FILE: {result['DESIGN_FILE']}")
            print(f"FEATURE_SPEC: {result['FEATURE_SPEC']}")
            print(f"FEATURE_DIR: {result['FEATURE_DIR']}")
            print(f"FEATURE_NAME: {result['FEATURE_NAME']}")


if __name__ == '__main__':
    main()

File: .roo/tools/setup-design.ts
/**
 * setup_design - Initialize design documentation for a feature or global design system
 *
 * Creates design documentation either for a specific feature or for the
 * global design system. Uses templates when available.
 *
 * Ported from: .zo/scripts/python/setup-design.py
 *
 * @example
 * ```typescript
 * // Setup feature design
 * const result = await setup_design({ json: true })
 * // Returns: {"MODE":"feature","DESIGN_FILE":"/path/to/design.md",...}
 *
 * // Setup global design system
 * const result = await setup_design({ json: true, globalMode: true })
 * // Returns: {"MODE":"global","DESIGN_FILE":"/path/to/.zo/design-system.md"}
 * ```
 *
 * @param json - Output results in JSON format (default: false)
 * @param globalMode - Setup global design system at .zo/design-system.md (default: false)
 * @param featureDir - Optional feature directory path (defaults to current branch)
 *
 * @returns JSON object or formatted text with design file paths
 *
 * @throws {Error} If execution environment validation fails
 * @throws {Error} If feature directory cannot be determined (in feature mode)
 *
 * @see {@link ../lib/util.ts} for path resolution utilities
 */

import { defineCustomTool } from '@roo-code/types';
import { parametersSchema as z } from '@roo-code/types';
import { existsSync, readFileSync, writeFileSync } from 'node:fs';
import { join, resolve } from 'node:path';
import {
  checkFileExists,
  getCurrentBranch,
  getFeaturePaths,
  getRepoRoot,
  resolvePath,
  validateExecutionEnvironment,
} from './lib/util.js';

export default defineCustomTool({
  name: 'setup_design',

  description: `
    Initialize design documentation for a feature or global design system.
    Creates design.md in the feature directory or .zo/design-system.md for global mode.
    Uses design templates when available.
  `,

  parameters: z.object({
    json: z
      .boolean()
      .optional()
      .describe('Output results in JSON format (compact)'),
    globalMode: z
      .boolean()
      .optional()
      .describe('Setup global design system (.zo/design-system.md)'),
    featureDir: z
      .string()
      .optional()
      .describe('Optional feature directory path (defaults to current branch)'),
  }),

  async execute({ json = false, globalMode = false, featureDir }, _context) {
    try {
      // Validate execution environment
      if (!validateExecutionEnvironment()) {
        const errorMsg = 'ERROR: Execution environment validation failed.';
        return json
          ? JSON.stringify({ error: errorMsg })
          : errorMsg;
      }

      const repoRoot = getRepoRoot();

      // Handle global vs feature mode
      if (globalMode) {
        return setupGlobalDesign(repoRoot, json);
      }

      return setupFeatureDesign(repoRoot, featureDir || '', json);

    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      return json
        ? JSON.stringify({ error: errorMessage })
        : `ERROR: ${errorMessage}`;
    }
  },
});

/**
 * Setup global design system
 */
function setupGlobalDesign(repoRoot: string, jsonMode: boolean): string {
  const designFile = join(repoRoot, '.zo', 'design-system.md');
  const featureSpec = '';
  const featureDir = repoRoot;
  const featureName = 'global';

  // Use global design system template if available
  const templatePath = join(repoRoot, '.zo', 'templates', 'design-system-template.md');

  if (!checkFileExists(designFile) && checkFileExists(templatePath)) {
    const content = readFileSync(templatePath, 'utf-8');
    writeFileSync(designFile, content, 'utf-8');
  }

  const result = {
    MODE: 'global',
    DESIGN_FILE: designFile,
    FEATURE_SPEC: featureSpec,
    FEATURE_DIR: featureDir,
    FEATURE_NAME: featureName,
  };

  if (jsonMode) {
    return JSON.stringify(result);
  }

  return [
    `MODE: ${result.MODE}`,
    `DESIGN_FILE: ${result.DESIGN_FILE}`,
  ].join('\n');
}

/**
 * Setup feature design
 */
function setupFeatureDesign(repoRoot: string, featureArg: string, jsonMode: boolean): string {
  const currentBranch = getCurrentBranch();

  // Handle optional feature directory argument
  let featureDir: string;
  let featureSpec: string;
  let featureName: string;

  if (featureArg) {
    // User provided a feature directory
    if (existsSync(featureArg)) {
      // Absolute or relative path provided
      featureDir = resolve(featureArg);
      featureSpec = join(featureDir, 'spec.md');
      featureName = featureDir.split('/').pop() || currentBranch;
    } else if (existsSync(join(repoRoot, featureArg))) {
      // Path relative to repo root
      featureDir = resolve(join(repoRoot, featureArg));
      featureSpec = join(featureDir, 'spec.md');
      featureName = featureDir.split('/').pop() || currentBranch;
    } else {
      const errorMsg = `Directory '${featureArg}' not found.`;
      return jsonMode
        ? JSON.stringify({ error: errorMsg })
        : errorMsg;
    }
  } else {
    // Auto-detect context from current branch
    const branchMatch = currentBranch.match(/^\d{3}-/);
    if (branchMatch) {
      // Branch name is the feature name
      const paths = getFeaturePaths();
      featureDir = paths.FEATURE_DIR;
      featureSpec = paths.FEATURE_SPEC;
      featureName = currentBranch;
    } else {
      // Fallback to current directory if valid structure
      if (checkFileExists('spec.md')) {
        featureDir = process.cwd();
        featureSpec = join(featureDir, 'spec.md');
        featureName = featureDir.split('/').pop() || currentBranch;
      } else {
        const errorMsg = [
          'Could not determine feature context.',
          'Run inside a feature branch or specify directory.',
        ].join(' ');
        return jsonMode
          ? JSON.stringify({ error: errorMsg })
          : errorMsg;
      }
    }
  }

  // Define output file for feature design
  const designFile = join(featureDir, 'design.md');

  // Use feature design template if available
  const templatePath = join(repoRoot, '.zo', 'templates', 'design-template.md');

  if (!checkFileExists(designFile) && checkFileExists(templatePath)) {
    let content = readFileSync(templatePath, 'utf-8');
    // Replace basic placeholders
    content = content.replace('[FEATURE NAME]', featureName);
    writeFileSync(designFile, content, 'utf-8');
  }

  const result = {
    MODE: 'feature',
    DESIGN_FILE: designFile,
    FEATURE_SPEC: featureSpec,
    FEATURE_DIR: featureDir,
    FEATURE_NAME: featureName,
  };

  if (jsonMode) {
    return JSON.stringify(result);
  }

  return [
    `MODE: ${result.MODE}`,
    `DESIGN_FILE: ${result.DESIGN_FILE}`,
    `FEATURE_SPEC: ${result.FEATURE_SPEC}`,
    `FEATURE_DIR: ${result.FEATURE_DIR}`,
    `FEATURE_NAME: ${result.FEATURE_NAME}`,
  ].join('\n');
}
